---
title: "SLE777_AS4_KS"
author: "Karmanpreet Sukhija"
date: "2025-09-28"
output: html_document
theme: cosmo
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Part 1

**Question 1**
First, the gene_expression.tsv file is imported, using the download.file command and read into R as a table. For the read.table command, stringsAsFactors was set to FALSE to ensure strings were imported as strings, not factors (Xie, 2020). This table is assigned to the variable gene_exp. The data is then inspected using the head command which provides the first 6 rows of the table. 
```{r}
URL <- "https://raw.githubusercontent.com/ghazkha/Assessment4/refs/heads/main/gene_expression.tsv"
NAME <- "gene_expression.tsv"
download.file(URL, destfile=NAME)
gene_exp <- read.table("gene_expression.tsv", stringsAsFactors = FALSE, header = TRUE, row.names = 1)
head(gene_exp)
```

**Question 2**
The mean gene count is then calculated by calculating the row mean, as each column is associated with a different sample and each row corresponds to a different gene. Therefore, the rowMeans function is used to calculate the mean gene count for each gene. The mean gene count is then stored under a new column in the table, with the name GENE.MEAN. The mean gene count for the first 6 genes is then observed using the head command. 
```{r}
gene_exp$GENE.MEAN <- rowMeans(gene_exp)
head(gene_exp)
```

**Question 3**
The table is then ordered, based on mean gene count, in descending order. This is conducted using the order command and the now sorted table is stored under the variable sorted_mean_genes. Following from this, the top 10 genes with the highest mean gene count is assessed using the head command, where n=10, to show the first 10 rows. 
```{r}
sorted_mean_genes <- gene_exp[order(gene_exp$GENE.MEAN, decreasing = TRUE),]
head(sorted_mean_genes, n=10)
```

**Question 4**
To determine the number of genes that have a mean gene count of less than 10, the data is subsetted using the subset command, and all the genes which have a mean gene count <10 are stored in the variable low_gene_number as a new table. The number of rows is then calculated to compute the number of genes present in this table. 
```{r}
low_gene_number <- subset(gene_exp, GENE.MEAN < 10)
nrow(low_gene_number)
```


**Question 5** 
The distribution of the mean gene count is then visualised using a histogram. As the data is heavily skewed, a second histogram is also generated, with a limit on the x-axis at 5000 and with more breaks (tubby and zx8754, 2015), to observe where most of the data is distributed and provide a better visualisation of the mean gene expression.
```{r}
hist(gene_exp$GENE.MEAN, 
     xlab = "Mean Gene Count", 
     ylab = "Frequency", 
     main = "Mean Gene Count", 
     col = "pink")

#More 'zoomed in' view of the histogram, as it is heavily skewed.
hist(gene_exp$GENE.MEAN, 
     xlab = "Mean Gene Count", 
     ylab = "Frequency", 
     main = "Mean Gene Count (Zoomed in 0-5000)", 
     col = "pink",
     xlim = c(0,5000),
     breaks = 500)

```

**Question 6**
Using the download.file command, the growth_data.csv file is imported. The column names are then investigated using the head command, which prints the fist 6 rows of the table. 
```{r}
URL2 = "https://raw.githubusercontent.com/ghazkha/Assessment4/refs/heads/main/growth_data.csv"
NAME2 = "growth_data.csv"
download.file(URL2, destfile = NAME2)
growth_data <- read.table("growth_data.csv", sep = ",", header = TRUE)
head(growth_data)
```
**Question 7**
The trees at the North-East and South-West sites are subsetted into two different variables. From there, the mean and standard deviations of the tree circumferences are calculated at the start of the study, in 2005, and at the end of the study, in 2020, at each site. 
```{r}
#North-East Site Trees
northeast_trees <- subset(growth_data, Site == "northeast")

#Start of the study
mean(northeast_trees$Circumf_2005_cm)
sd(northeast_trees$Circumf_2005_cm)

#End of the study
mean(northeast_trees$Circumf_2020_cm)
sd(northeast_trees$Circumf_2020_cm)

#--------------------------------------------------------------------#

#South-West Site Trees
southwest_trees <- subset(growth_data, Site == "southwest")

#Start of the study
mean(southwest_trees$Circumf_2005_cm)
sd(southwest_trees$Circumf_2005_cm)

#End of the study 
mean(southwest_trees$Circumf_2020_cm)
sd(southwest_trees$Circumf_2020_cm)
```

**Question 8**
A boxplot is then generated to visualise the distribution of tree circumference at the start of the study and at the end of the study, at each site to then compare the differences between the sites. 
```{r}
#Boxplot of tree circumference at each site at the start of the study, 2005
boxplot(northeast_trees$Circumf_2005_cm, southwest_trees$Circumf_2005_cm, 
        ylab = "Tree Circumference (cm)", 
        xlab = "Site",
        main = "Tree Circumference (cm) in 2005 at Northeast and Southwest Sites", 
        names = c("North-East", "South-West"))

#Boxplot of tree circumference at each site at the end of the study, 2020
boxplot(northeast_trees$Circumf_2020_cm, southwest_trees$Circumf_2020_cm, 
        ylab = "Tree Circumference (cm)", 
        xlab = "Site",
        main = "Tree Circumference (cm) in 2020 at Northeast and Southwest Sites", 
        names = c("North-East", "South-West"))
```

**Question 9**
The mean growth over the last 10 years (2010 - 2020) is calculated through the subtraction of the tree circumference in 2010 from the tree circumference in 2020. The mean differences between the two timepoints is then calculated using the mean function, for both sites. 
```{r}
#Mean 10 year growth for North-East site
mean(northeast_trees$Circumf_2020_cm - northeast_trees$Circumf_2010_cm)

#Mean 10 year growth for South-West site
mean(southwest_trees$Circumf_2020_cm - southwest_trees$Circumf_2010_cm)
```

**Question 10**
Welch's t-test is then used to determine whether the 10 year growth at the North-East site is statitsically significantly different to the 10 year growth at the South-West site. This is performed by first generating new columns in the tables to store the values of the 10 year growth of tree circumferences at each site. 

To determine whether to perform a Students t-test of Welch's t-test, an F-test is conducted to test for equal variance (O’Brien, 2023). 
```{r}
#10 year growth at the North-East site
northeast_trees$Tenyear_growth <- northeast_trees$Circumf_2020_cm - northeast_trees$Circumf_2010_cm

#10 year growth at the South-West site
southwest_trees$Tenyear_growth <- southwest_trees$Circumf_2020_cm - southwest_trees$Circumf_2010_cm

#Determine the appropriate t-test, Welch's vs Students, via F-test
var.test(northeast_trees$Tenyear_growth, southwest_trees$Tenyear_growth)

```

As the F-test generates a statistically significant p-value, this means that the two variables do not have equal variance and thus Welch's t-test is most appropriate to use. These variables are then used to conduct Welch's t-test, which does not assume equal standard deviations between the two variables.  
```{r}
#Welch's t-test 
t.test(northeast_trees$Tenyear_growth, southwest_trees$Tenyear_growth)
```


# Part 2

**Question 1**
First, the packages seqinr and R.utils are loaded.
```{r}
suppressPackageStartupMessages({
  library("seqinr") 
  library("R.utils") 
})
```

The download.file command is used to download the FASTA files of the coding sequences of E.coli and S.thermophilus. The gunzip command is then used to unzip the files. Following this, the seqinr package is used to read-in the FASTA files into a FASTA file format. 
```{r}
#Downloading and reading in the E.coli coding sequence
URL_e.coli="http://ftp.ensemblgenomes.org/pub/bacteria/release-53/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655_gca_000005845/cds/Escherichia_coli_str_k_12_substr_mg1655_gca_000005845.ASM584v2.cds.all.fa.gz"
download.file(URL_e.coli, destfile = "ecoli_cds.fa.gz")
gunzip("ecoli_cds.fa.gz")

cds_ecoli <- seqinr::read.fasta("ecoli_cds.fa")

#Downloading and reading in the S.thermophilus coding sequence
URL_s.thermo = "https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-62/fasta/bacteria_121_collection/streptococcus_thermophilus_cag_236_gca_000434755/cds/Streptococcus_thermophilus_cag_236_gca_000434755.MGS236.cds.all.fa.gz"
download.file(URL_s.thermo, destfile = "sthermo_cds.fa.gz")
gunzip("sthermo_cds.fa.gz")

cds_sthermo <- seqinr::read.fasta("sthermo_cds.fa")

list.files()
```

The calculate the number of coding sequences present in each organism, the length command is used which returns the number of coding sequences present. The head command is also used to get a feel of the data and ensure it is in the correct format. 
```{r}
#Number of coding sequences present in E.coli and summary of the imported data
length(cds_ecoli)
head(summary(cds_ecoli))

#Number of coding sequences present in S.thermophilus and summary of the imported data
length(cds_sthermo)
head(summary(cds_sthermo))
```

**Question 2**
To compute the number of base pairs present in the total coding sequence, the length of all coding sequences are summed by storing the lengths of each coding sequence under the len_ecoli and len_sthermo variables. Once these variables have been summed, the output results in the total number of base pairs present in the coding sequence of each organism.
```{r}
#Total number of base pairs present in the coding sequence of E.coli
len_ecoli <- as.numeric(summary(cds_ecoli)[,1])
sum(len_ecoli)

#Total number of base pairs present in the coding sequence of S.thermophilus
len_sthermo <- as.numeric(summary(cds_sthermo)[,1])
sum(len_sthermo)
```

**Question 3**
To compare the distribution of coding sequence length, a boxplot is used to visualise this parameter. Following this, the mean and median functions are used to calculate the mean and median coding sequence length in each organism.
```{r}
#Visualisation of distribution of coding sequence length in each organism 
boxplot(len_ecoli, len_sthermo, 
        ylab = "Coding Sequence Length (bp)", 
        main = "Coding Sequence Lengths", 
        names = c("E.coli", "S.thermophilus"))

#Calculation of mean and median coding sequence length in E.coli
mean(len_ecoli)
mean(len_sthermo)

#Calculation of mean and median coding sequence length in S.thermophilus
median(len_ecoli)
median(len_sthermo)
```

**Question 4**
As the current data is present as a dataframe, in order to allow for the DNA composition to be calculated, the coding sequence is unlisted so that it is now a long string of nucleotides. This allows for the number of nucleotides to be present and thus the nucleotide composition of the total coding sequence to be visualised, using a barplot, for each organism. 
```{r}
#Unlisting of E.coli coding sequence
dna_ecoli <- unlist(cds_ecoli)
#Counting the total number of each nucleotide present in the total coding sequence of E.coli
dna_comp_ecoli <- count(dna_ecoli,1)
dna_comp_ecoli
#Visualisation of nucleotide composition of total coding sequence of E.coli
barplot(dna_comp_ecoli, 
        xlab = "Nucleotides", 
        ylab = "Frequency", 
        main = "E.coli Coding Sequence Nucleotide Composition")

#Unlisting of S.thermophilus coding sequence
dna_sthermo <- unlist(cds_sthermo)
#Counting the total number of each nucleotide present in the total coding sequence of S.thermophilus
dna_comp_sthermo <- count(dna_sthermo,1)
dna_comp_sthermo
#Visualisation of nucleotide composition of total coding sequence of S.thermophilus
barplot(dna_comp_sthermo, 
        xlab = "Nucleotides", 
        ylab = "Frequency", 
        main = "S.thermophilus Coding Sequence Nucleotide Composition")
```

The same is then done for the total protein sequence, which is first translated using the seqinr package. Following this, the alphabet which corresponds to the amino acids present in E.coli and S.thermophilus is stored and used to count the number of each amino acid present in the coding sequence of each organism. The amino acid composition of the total coding sequence of each organism is then visualised using a barplot.
```{r}
#Translation of full DNA coding sequence of E.coli to amino acid coding sequence
fullprotein_ecoli <- unlist(lapply(cds_ecoli, translate))

#All unique amino acids present within E.coli are stored, expect for * which corresponds to a stop codon. 
aa <- unique(fullprotein_ecoli)
aa <- aa[aa != "*"]

#The number of amino acids present in the coding sequence of E.coli is counted 
protein_comp_ecoli <- count(fullprotein_ecoli,wordsize=1,alphabet=aa)
protein_comp_ecoli

#The amino acid composition of the total coding sequence of E.coli is visualised using a barplot, with the x-axis labels being shrunk to include all amino acids on the axis
barplot(protein_comp_ecoli,
        xlab = "Amino Acids",
        ylab = "Frequency", 
        main = "E.coli Amino Acid Composition",
        cex.names = 0.7)

#######

#Translation of full DNA coding sequence of S.thermophilus to amino acid coding sequence
fullprotein_sthermo <- unlist(lapply(cds_sthermo, translate))

#The number of amino acids present in the coding sequence of S.thermophilus is counted, using the same alphabet as used with E.coli
protein_comp_sthermo <- count(fullprotein_sthermo,wordsize=1,alphabet=aa)
protein_comp_sthermo

#The amino acid composition of the total coding sequence of S.thermophilus is visualised using a barplot, with the x-axis labels being shrunk to include all amino acids on the axis
barplot(protein_comp_sthermo,
        xlab = "Amino Acids",
        ylab = "Frequency", 
        main = "S.thermophilus Amino Acid Composition",
        cex.names = 0.7)
```

**Question 5**
Using the seqinr package, the codon usage indices are calculated. The focus here is on the Relative Synonymous Codon Usage (RSCU) values, which corresponds to how preferred a certain codon is for a particular amino acid, compared to other codons which are used to code for that amino acid. The RSCU value is how codon usage bias is calculated within this assessment. The RSCU values are also visualised for each codon. 
```{r}
#Calculation of codon usage indices for E.coli
codonuse_ecoli <- uco(unlist(cds_ecoli), index = "rscu", as.data.frame = TRUE)
codonuse_ecoli

#Visualisation of RSCU values of each codon present in the coding sequence of E.coli
barplot(codonuse_ecoli$RSCU, 
        names.arg = codonuse_ecoli$codon, 
        las = 2, 
        xlab = "Codon", 
        ylab = "RSCU Value", 
        main = "Codon Usage Bias in E.coli", 
        col = "powderblue",
        cex.names = 0.5)

#Calculation of codon usage indices for S.thermophilus
codonuse_sthermo <- uco(unlist(cds_sthermo), index = "rscu", as.data.frame = TRUE)
codonuse_sthermo

#Visualisation of RSCU values of each codon present in the coding sequence of S.thermophilus
barplot(codonuse_sthermo$RSCU, 
        names.arg = codonuse_sthermo$codon, 
        las = 2, 
        xlab = "Codon", 
        ylab = "RSCU Value", 
        main = "Codon Usage Bias in S.thermophilus", 
        col = "powderblue",
        cex.names = 0.5)
```


**Question 6**
To compute the most under- and over-represented k-mers of length 3-5, the count command is used to count characters of length 3-5, using the alphabet previously described when looking at the amino acid composition of the coding sequence of each organism. The k-mer counts are then sorted in ascending order to view the top 10 most under- and over-represented sequences. Following this, the sorted k-mer counts are visualised using the barplot function to highlight the varying frequency of each k-mer. 
```{r}
#Computation of k-mers of length 3 in E.coli
ecoli_counts3 <- count(fullprotein_ecoli, wordsize = 3, alphabet = aa)
#Sorting k-mer list in ascending order
sorted_ecoli_counts3 <- ecoli_counts3[order(ecoli_counts3)]
#Returning the top 10 most under- and over-represented k-mers of length 3 in E.coli
head(sorted_ecoli_counts3, n=10)
tail(sorted_ecoli_counts3, n=10)

#Visualisation of sorted k-mer count as a barplot
barplot(sorted_ecoli_counts3,
        las = 2, 
        xlab = "K-mer", 
        ylab = "Frequency", 
        main = "K-mers of Length 3 Present in E.coli", 
        cex.names = 0.5)

#The above is repeated for k-mers of length 4 and 5 

ecoli_counts4 <- count(fullprotein_ecoli, wordsize = 4, alphabet = aa)
sorted_ecoli_counts4 <- ecoli_counts4[order(ecoli_counts4)]
head(sorted_ecoli_counts4, n=10)
tail(sorted_ecoli_counts4, n=10)

barplot(sorted_ecoli_counts4,
        las = 2, 
        xlab = "K-mer", 
        ylab = "Frequency", 
        main = "K-mers of Length 4 Present in E.coli", 
        cex.names = 0.5)

ecoli_counts5 <- count(fullprotein_ecoli, wordsize = 5, alphabet = aa)
sorted_ecoli_counts5 <- ecoli_counts5[order(ecoli_counts5)]
head(sorted_ecoli_counts5, n=10)
tail(sorted_ecoli_counts5, n=10)

barplot(sorted_ecoli_counts5,
        las = 2, 
        xlab = "K-mer", 
        ylab = "Frequency", 
        main = "K-mers of Length 5 Present in E.coli", 
        cex.names = 0.5)

##############################################

#k-mers of length 3-5 are computed for S.thermophilus, same as what was done for E.coli

sthermo_counts3 <- count(fullprotein_sthermo, wordsize = 3, alphabet = aa)
sorted_sthermo_counts3 <- sthermo_counts3[order(sthermo_counts3)]
head(sorted_sthermo_counts3, n=10)
tail(sorted_sthermo_counts3, n=10)

barplot(sorted_sthermo_counts3,
        las = 2, 
        xlab = "K-mer", 
        ylab = "Frequency", 
        main = "K-mers of Length 3 Present in S.thermophilus", 
        cex.names = 0.5)

sthermo_counts4 <- count(fullprotein_sthermo, wordsize = 4, alphabet = aa)
sorted_sthermo_counts4 <- sthermo_counts4[order(sthermo_counts4)]
head(sorted_sthermo_counts4, n=10)
tail(sorted_sthermo_counts4, n=10)

barplot(sorted_sthermo_counts4,
        las = 2, 
        xlab = "K-mer", 
        ylab = "Frequency", 
        main = "K-mers of Length 4 Present in S.thermophilus", 
        cex.names = 0.5)

sthermo_counts5 <- count(fullprotein_sthermo, wordsize = 5, alphabet = aa)
sorted_sthermo_counts5 <- sthermo_counts5[order(sthermo_counts5)]
head(sorted_sthermo_counts5, n=10)
tail(sorted_sthermo_counts5, n=10)

barplot(sorted_sthermo_counts5,
        las = 2, 
        xlab = "K-mer", 
        ylab = "Frequency", 
        main = "K-mers of Length 5 Present in S.thermophilus", 
        cex.names = 0.5)

```

# Session Info 
```{r}
sessionInfo()
```


# References 
Advanced R barplot customization [Online]. Available: https://r-graph-gallery.com/210-custom-barplot-layout.html [Accessed Oct 2025].
Chapter 11 Histogram | Basic R Guide for NSC Statistics [Online]. Available: https://bookdown.org/dli/rguide/histogram.html [Accessed Oct 2025].
O’BRIEN, K. 2023. Testing for Equality of Variance [Online]. Available: https://rpubs.com/DragonflyStats/Variance-Test [Accessed Oct 2025].
TUBBY & ZX8754. 2015. Change scale of histogram in R [Online]. stack overflow. Available: https://stackoverflow.com/questions/30378955/change-scale-of-histogram-in-r [Accessed Oct 2025].
WHV20 & ISSAC. 2015. names on the x-axis of barplot in R? [Online]. stack overflow. Available: https://stackoverflow.com/questions/28416992/names-on-the-x-axis-of-barplot-in-r [Accessed Oct 2025].
XIE, Y. 2020. stringsAsFactors. Available from: https://blog.r-project.org/2020/02/16/stringsasfactors/ [Accessed Oct 2025].
